pkg/operator/reports.go

```diff
@@ -734,6 +745,59 @@ func (op *Reporting) runReport(logger log.FieldLogger, report *metering.Report)
 	return nil
 }
 
+// We check first for Reports depending on this Report, return if any, and then check ReportQueries depending on Report
+func isReportNotUsedAsInput(report *metering.Report, op *Reporting) bool {
+	// Consider Reports referencing this report in the namespace of this report
+	reports, err := op.reportLister.Reports(report.Namespace).List(labels.Everything())
+	if err != nil {
+		op.logger.Errorf("unable to determine list of reports that might be input for report: %s", report.Name)
+		return false
+	}
+	for _, report := range reports {
+		// If Report has no spec.inputs it can't be depending on this report.
+		if report.Spec.Inputs == nil {
+			continue
+		}
+
+		if op.doReportsDependOnThisReport(report, reports) {
+			op.logger.Infof("reports exist that use report %s as input, will not delete though retention period has expired", report.Name)
+			return false
+		}
+	}
+	// Consider ReportQueries referencing this report in the namespace of this report
+	reportQueries, err := op.reportQueryLister.ReportQueries(report.Namespace).List(labels.Everything())
+	if err != nil {
+		op.logger.Errorf("unable to determine list of ReportQueries that might be input for report: %s", report.Name)
+		return false
+	}
+	for _, reportQuery := range reportQueries {
+		// If ReportQuery has no spec.inputs it can't be depending on this report.
+		if reportQuery.Spec.Inputs == nil {
+			continue
+		}
+
+		if op.reportQueriesDependOnThisReport(reportQuery) {
+			op.logger.Infof("some ReportQueries exist that use report %s as input, will not delete though retention period has expired", report.Name)
+			return false
+		}
```

Comment:
Just thinking out loud here with roll-up reports in mind again. Reading this code again, it appears that if we automatically purge the monthly Report as the user-provided retention period has been met and that Report's inputs reference a weekly Report who also has a retention period configured. When we iterate over the list of Reports in the reporting-operator namespace, we'll see that it doesn't have any dependencies, but when we iterate of the list of ReportQueries, we would see one of those custom resources use that Report as an input, and we would mark the weekly Report as "expired, but has dependencies so not deleting". Is this right?

ading this code is what happens if we automatically purge the Report (monthly) whose inputs references another Report (weekly), we delete the monthyl Report that references that other Report, but we can't delete that other

Reply:
If so, that seems like bad UX, at least for the roll-up use-case.

Reply:
I got confused there, probably due to my lack of familiarity with roll-ups. I read rollup-reports.md, but will have to read it again to see what the answer is here...

Reply:
ok so reading and thinking, it seems like the fix here is a follow-on task where we add retention times to ReportQuery so that they can be set to allow deletion of the sub-reports in the roll-up scenario?

Reply:
Let's keep this implementation as-is for now, with the intention of documenting this limitation w.r.t. roll-up Reports + user-provided retention period.